finat/tensorfiniteelement.py

```diff
@@ -63,15 +66,18 @@ def entity_dofs(self):
         raise NotImplementedError("No one uses this!")
 
     def space_dimension(self):
-        return numpy.prod((self._base_element.space_dimension(),) + self._shape)
+        return int(numpy.prod(self.index_shape))
 
     @property
     def index_shape(self):
-        return self._base_element.index_shape + self._shape
+        if self._transpose:
+            return self._shape + self._base_element.index_shape
+        else:
+            return self._base_element.index_shape + self._shape
 
     @property
     def value_shape(self):
-        return self._base_element.value_shape + self._shape
+        return self._shape + self._base_element.value_shape
```

Comment:
This is relevant in the rare case when one takes the vector of naturally vector-valued element, e.g. `VectorElement('BDM', triangle, 2)`. In this case UFL defines what the value shape should be like, and UFL says:

    In [7]: VectorElement('BDM', triangle, 2, dim=3).value_shape()
    Out[7]: (3, 2)
    
    In [8]: VectorElement('BDM', triangle, 2, dim=3).reference_value_shape()
    Out[8]: (3, 2)

This is the same, irrespective of the DoF ordering.